File: 977-distinct-subsequences-ii/distinct-subsequences-ii.java
class Solution {
    public int distinctSubseqII(String s) {
    int mod = 1_000_000_007;
    int ans = 0;
    int[] dp = new int[26];
    
    for (int i = 0; i < s.length(); i++) {
        int c = s.charAt(i) - 'a';
        int add = (ans + 1 - dp[c] + mod) % mod;
        ans = (ans + add) % mod;
        dp[c] = (dp[c] + add) % mod;
    }
    return ans;
    }
}

File: 977-distinct-subsequences-ii/distinct-subsequences-ii.py
class Solution(object):
    mod=10**9+7
    def distinctSubseqII(self, s):
        """
        :type s: str
        :rtype: int
        """
        ans=0
        dp=[0]*26
        for i in s:
            c=ord(i)-97
            new=ans+1-dp[c]
            ans=(ans+new)%self.mod
            dp[c]=(dp[c]+new)%self.mod
        #print(dp)
        return ans